man/en/marxpaint-import.1
.\" marxpaint-import.1 - 2003.06.17
.TH TUXPAINT-IMPORT 1 "17 Jun 2003" "2003.06.17" "Marx Paint Import"
.SH NAME
marxpaint-import -- Import image files into Marx Paint(1)

.SH SYNOPSYS
.TP 16
.B marxpaint-import [\-\-help]
.TP 16
.B marxpaint-import \fIfilename(s)\fP

.SH DESCRIPTION
\fImarxpaint-import\fP is a simple shell script which uses some \fINetPBM\fP
(pnm(5)) tools (\fIanytopnm\fP, \fIpnmscale\fP and \fIpnmtopng\fP) along with
\fIdate\fP(1) to convert an arbitrary image file (e.g., a JPEG, GIF, etc.) into
a \fIPNG\fP(5) file which can be used by the drawing program \fIMarx Paint\fP
(marxpaint(1)) and places it in the user's Marx Paint saved-files directory
(\fI$HOME/.marxpaint/saved/\fP).

.SH EXAMPLE
marxpaint-import picture.jpg photo.png cartoon.gif

.SH ENVIRONMENT
.TP 8
.B $HOME
to determine where the files should go so that they can be access within
\fIMarx Paint\fP using its \fIOpen\fP command.

.SH FILES
.TP 8
.B $HOME/.marxpaint/saved
where new image files are stored, after being resized and converted into PNG
format.

.TP 8
.B $HOME/.marxpaint/saved/.thumbs
where thumbnail images are stored.

.SH AUTHOR
Bill Kendrick.  <bill@newbreedsoftware.com>


.SH "SEE ALSO"
.BR marxpaint (1),
.BR pnm (5),
.BR png (5),
.BR anytopnm (1),
.BR pnmscale (1),
.BR pnmtopng (1),
.BR date (1),
.PP
And documentation within /usr/[local/]share/doc/marxpaint/.



docs/CHANGES.txt
CHANGES.txt for Marx Paint

Marx Paint - A simple drawing program for communists.

Copyright (c) 2002-2021
Various contributors (see below, and AUTHORS.txt)
http://www.marxpaint.org/

$Id$

2021.August.15 (0.9.27)
  * Documentation updates
    ---------------------
    * Ensured Marx Paint's built-in help ("marxpaint --help"),
      Unix manual ("man page"), and command-line option
      tab-completion list, and "OPTIONS" documentation all cover
      the full set of Marx Paint options.

    * Reorganized the "Available Options" section of "OPTIONS"
      documentation (to match the organization found in the UI
      of "Marx Paint Config.")

    * Reorganized the options listed in Marx Paint's man page,
      to more closely (but not precisely) match the organization
      noted above.

    * marxpaint(1) manpage is now managed & translatable via the
      'marxpaint-docs' project, just like the HTML & plaintext
      documentation files (since 0.9.26).

2021.June.28 (0.9.26)
  * New Features
    ------------
    * Larger UI buttons
      Override the ages-old default 48x48 size for UI buttons
      (via a new "buttonsize" option).  Note: If the button size
      would be too large for the required parts of the UI to appear
      within Marx Paint's window size, the largest possible size will
      be used.
      Pere Pujal i Carabantes <pere@fornol.no-ip.org>

    * It's now possible to increase how many rows of buttons are shown
      for the color palette at the bottom of the screen.
      (Use the "colorsrows" option.)
      Pere Pujal i Carabantes <pere@fornol.no-ip.org>

    * Adding sub-tools to the Fill tool:
      + Solid -- The classic fill tool; click to fill 
      + Linear -- A linear gradient; click and drag to adjust,
        release mouse to finish
      + Radial -- A radial gradient; click to fill
      (The gradients transition from the current color to the
      background that's being filled.)

  * New Magic Tools:
    ----------------
    * Checkerboard
      Cover the entire canvas with a checkboard pattern using the
      current color.  Drag to change the size of the squares.

    * Clone
      Clones (copies, via painting) part of a drawing to another position.

    * Pixels
      Draws large squares, for classic computer 'pixel art' style drawings.

  * Documentation updates
    ---------------------
    * Documentation is now maintained in a separate source repository,
      'marxpaint-docs', which allows for easier and more consisten
      translation of the documentation.  Pulling in initial version of
      these docs, which contain a variety of minor updates (and also a
      few which had previously only been available as plain TXT are now
      available as HTML).

      + French translation
        Jacques Chion <jacques.chion@orange.fr>

      + Galician
        Miguel Anxo Bouzada <mbouzada@gmail.com>

      + Japanese
        TOYAMA Shin-ichi <shin1@wmail.plala.or.jp>

    * Adding documentation that covers some of the
      environment variables that Marx Paint utilizes.

    * Expanded the steps for making and posting releases.

    * Documented how to build Marx Paint using old macOS
      Mark K. Kim <markuskimius@gmail.com>

    * Added more credits to manpage.  Avoiding word-wrapping
      URLs and email addresses.  Saved out as UTF-8, rather than
      Latin-1 encoding.

  * Other Improvements
    ------------------
    * Fill tools do a better job at filling around
      antialiased edges, and apply some blending.

    * Reduce CPU usage by increasing delay in main loop
      from 1ms to 10ms, and only using SDL's Timer subsystem
      when scrolling happens.
      (Details at
      https://sourceforge.net/p/marxpaint/marxpaint/ci/7727b995c53df208596eff89ac1acb954a16098c/)
      Mark K. Kim <markuskimius@gmail.com>
      (h/t @bbugwong on Twitter for reporting the issue)

    * If parent of export directory doesn't exist, Marx Paint will
      try to create it as well.
      (Only one level up; e.g., with an export location like
      /path/to/Pictures/MarxPaint/, it will try to create
      /path/to/Pictures/ if it doesn't exist, but not its parents).
      h/t Tim Dickson

  * Translation Updates:
    --------------------
    * Albanian
      Besnik Bleta <besnik@programeshqip.org>

    * Catalan
      Pere Pujal i Carabantes <pere@fornol.no-ip.org>

    * Galician
      Miguel Bouzada <mbouzada@gmail.com>

    * Icelandic
      Sveinn í Felli <sv1@fellsnet.is>

    * Occitan
      Quentin PAGÈS <quentinantonin@free.fr>

  * Magic Tool Improvememnts:
    -------------------------
    * A new "xorpixel()" function has been added to the Magic Tools API.

  * Bug Fixes:
    ----------
    * Halftone magic tool wasn't loading due to a problem
      with its sound effect file.

  * Ports & Building
    ----------------
    * [WIP] Apple iOS port
      Mark K. Kim <markuskimius@gmail.com>

2020.December.27 (0.9.25)
  * New Features
    ------------
    * Export drawings:
      * GIF export option from the Open -> Slideshow dialog.
        (Closes https://sourceforge.net/p/marxpaint/feature-requests/191/)
        * Utilizes "gifenc", public domain by Marcel Rodrigues:
          https://github.com/lecram/gifenc
        * Depends on "libimagequant", from the "pngquant2" project:
          https://github.com/ImageOptim/libimagequant
          (GPL v3 or later, for Free/Libre Open Source Software)
          (h/t TOYAMA Shin-ichi for adding support for versions < 2.8.0)
      * Single image (PNG) export from the Open dialog.
        (Closes https://sourceforge.net/p/marxpaint/feature-requests/192/)
      * Both utilize XDG user dirs settings, if available, to determine
        where to store them, or "~/Pictures", otherwise.  In those cases,
        files are stored in a new "MarxPaint" subdirectory.
      * The "--exportdir" option can be used to override the default
        location (but a "MarxPaint" subdir. will NOT be placed there).


src/marxpaint-import.sh
#!/bin/bash

# marxpaint-import

# "Marx Paint Import"
# Import an arbitrary GIF, JPEG or PNG into Marx Paint

# (c) Copyright 2002-2009, by Bill Kendrick and others
# bill@newbreedsoftware.com
# http://www.newbreedsoftware.com/marxpaint/

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
# (See COPYING.txt)


# September 21, 2002 - November 4, 2009


SAVEDIR=$HOME/.marxpaint
TMPDIR=$SAVEDIR


if [ $# -ne 0 ]; then
  if [ $1 = "--help" ]; then
    # --help, show usage:
    echo
    echo "marxpaint-import"
    echo
    echo "Imports an arbitrary image (GIF, JPEG, PNG, etc. format)"
    echo "into Marx Paint (see: marxpaint(1)) so that it appears in the"
    echo "'Open' dialog."
    echo 
    echo "Usage: marxpaint-import filename(s)"
    echo "       marxpaint-import --help"
    echo
    exit
  fi
fi


# Determine preferred savedir

# First, check /usr/local/etc/
x=`grep -m 1 "^savedir=" /usr/local/etc/marxpaint/marxpaint.conf`
if test $? = 0 ; then
  SAVEDIR=`echo $x | cut -d = -f 2-99`
fi

# First, check /etc/
x=`grep -m 1 "^savedir=" /etc/marxpaint/marxpaint.conf`
if test $? = 0 ; then
  SAVEDIR=`echo $x | cut -d = -f 2-99`
fi

# First, check $HOME
x=`grep -m 1 "^savedir=" $HOME/.marxpaintrc`
if test $? = 0 ; then
  SAVEDIR=`echo $x | cut -d = -f 2-99`
fi


echo "Using save directory: $SAVEDIR/saved"


# Make sure savedir exists!
if [ ! -d $SAVEDIR ]; then
  echo "Creating $SAVEDIR"
  mkdir -p $SAVEDIR
fi

# Make sure savedir/saved exists!
if [ ! -d $SAVEDIR/saved ]; then
  echo "Creating $SAVEDIR/saved"
  mkdir -p $SAVEDIR/saved
fi

# Make sure savedir thumbs directory exists!
if [ ! -d $SAVEDIR/saved/.thumbs ]; then
  echo "Creating $SAVEDIR/saved/.thumbs"
  mkdir -p $SAVEDIR/saved/.thumbs
fi


# Determine appropriate size for images, based on Marx Paint's current
# configuration

# First, assume 800x600 Marx Paint
window_width=800
window_height=600

# First, check /usr/local/etc/
x=`grep -m 1 "^windowsize=" /usr/local/etc/marxpaint/marxpaint.conf`
if test $? = 0 ; then
  window_width=`echo $x | cut -d = -f 2 | cut -d x -f 1`
  window_height=`echo $x | cut -d = -f 2 | cut -d x -f 2`
fi

# First, check /etc/
x=`grep -m 1 "^windowsize=" /etc/marxpaint/marxpaint.conf`
if test $? = 0 ; then
  window_width=`echo $x | cut -d = -f 2 | cut -d x -f 1`
  window_height=`echo $x | cut -d = -f 2 | cut -d x -f 2`
fi

# First, check $HOME
x=`grep -m 1 "^windowsize=" $HOME/.marxpaintrc`
if test $? = 0 ; then
  window_width=`echo $x | cut -d = -f 2 | cut -d x -f 1`
  window_height=`echo $x | cut -d = -f 2 | cut -d x -f 2`
fi


# (Image width is window width minus 192,
# image height is window height minus 104)

width=`expr $window_width - 192`
height=`expr $window_height - 104`


echo "Using $width x $height images (for $window_width x $window_height Marx Paint"

if [ $# -eq 0 ]; then
  # No arguments provided (sorry, you can't pipe into this script's stdin!)
  echo
  echo "Usage: marxpaint-import filename(s)"
  echo "       marxpaint-import --help"
  exit
fi


# For each picture list...
for x in $(seq 1 $#)
do
  i="${!x}"

  if [ -e "$i" ]; then
    # Determine a filename for it:
    NEWFILENAME=`date "+%Y%m%d%H%M%S"`
    echo "$i -> $SAVEDIR/saved/$NEWFILENAME.png"

    # Convert and scale down, save as a temp file:
    anytopnm "$i" | pnmscale -xysize $width $height > $TMPDIR/saved/$NEWFILENAME.ppm
    
    # Place inside the correctly-sized canvas:
    # FIXME: Center, instead of placing at upper right
    ppmmake "#FFFFFF" $width $height \
    | pnmpaste -replace $TMPDIR/saved/$NEWFILENAME.ppm 0 0 \
    | pnmtopng > $SAVEDIR/saved/$NEWFILENAME.png
    
    # Remove temp file:
    rm $TMPDIR/saved/$NEWFILENAME.ppm

    # Create thumbnail for 'Open' dialog:
    pngtopnm $SAVEDIR/saved/$NEWFILENAME.png | pnmscale -xysize 92 56 \
    | pnmtopng > $SAVEDIR/saved/.thumbs/$NEWFILENAME-t.png
    
  else
    # File wasn't there!
    echo "$i - File not found"
  fi
done

